File: web/src/app/core/log.ts
import { Injectable, inject } from '@angular/core';

import { Bridge } from './bridge';

/** The levels Rust knows how to read back. See `crates/mc-app/src/log.rs`. */
export type Level = 'error' | 'warn' | 'info' | 'debug' | 'trace';

/**
 * What the front tells Rust's log.
 *
 * ## Why not `console.log`
 *
 * The window has no inspector in production: what you write there goes into
 * the void, and it's in production that the defects we're looking for show
 * up. Lines therefore go to Rust, which writes them into ITS log — same
 * file, same clock, same order as its own.
 *
 * It's the only way to read a sequence that crosses two windows and a
 * process. Stitching two logs back together by hand requires knowing in
 * what order things happened, which is precisely the question being asked.
 *
 * ## It NEVER fails the caller
 *
 * Logging is a diagnostic operation: a lost line costs a line, an exception
 * thrown from a diagnostic `finally` costs the very action being observed.
 * Everything is swallowed.
 */
@Injectable({ providedIn: 'root' })
export class Log {
  private readonly bridge = inject(Bridge);

  /** A milestone in the sequence: what the window just did. */
  step(message: string): void {
    this.write('info', message);
  }

  /** A detail only read when someone is looking for it. */
  detail(message: string): void {
    this.write('debug', message);
  }

  /** Something unexpected, that doesn't prevent continuing. */
  concern(message: string): void {
    this.write('warn', message);
  }

  private write(level: Level, message: string): void {
    // Also to the console: in front of the dev server, that's where you
    // look, and the duplication costs nothing.
    console.info(`[${level}] ${message}`);
    void this.bridge.log(level, message).catch(() => {});
  }
}

File: crates/mc-app/src/log.rs
//! What the front end tells the log.
//!
//! ## Why the front end writes into Rust's log
//!
//! The window has no inspector in production: `console.log` goes into the
//! void there. But the bugs that matter are SEQUENCE bugs between two
//! windows and a process — "who navigated, when, and in which one" — and
//! they can only be read if both sides write into the same stream, in
//! order, with the same clock.
//!
//! Two separate logs that need to be stitched together by hand are worse
//! than one: the stitching error is precisely the error being sought.
//!
//! ## The label is NOT sent by the front end
//!
//! It's read from the calling window, which Tauri injects. It's the only
//! point in the setup where you can't get the window wrong — and getting
//! the window wrong is exactly the bug this module exists to track.

use tauri::Window;

/// What the front end may ask for, and nothing else.
///
/// A PURE function: it carries the module's only rule, and that's why it's
/// the one being tested. An unknown level maps to `info` rather than losing
/// the line — a diagnostic message that disappears because it was mislabeled
/// is the opposite of what's wanted here.
fn level_of(raw: &str) -> tracing::Level {
    match raw {
        "error" => tracing::Level::ERROR,
        "warn" => tracing::Level::WARN,
        "debug" => tracing::Level::DEBUG,
        "trace" => tracing::Level::TRACE,
        _ => tracing::Level::INFO,
    }
}

/// Writes a line from the front end, under its window's label.
///
/// The target is `front`: `RUST_LOG=front=trace` is then enough to see only
/// the front end, and `RUST_LOG=warn,front=trace` to see it alone in the
/// middle of a silent startup.
#[tauri::command]
pub fn log(window: Window, level: String, message: String) {
    let label = window.label().to_string();
    match level_of(&level) {
        tracing::Level::ERROR => {
            tracing::error!(target: "front", window = %label, "{message}")
        }
        tracing::Level::WARN => tracing::warn!(target: "front", window = %label, "{message}"),
        tracing::Level::DEBUG => {
            tracing::debug!(target: "front", window = %label, "{message}")
        }
        tracing::Level::TRACE => {
            tracing::trace!(target: "front", window = %label, "{message}")
        }
        tracing::Level::INFO => tracing::info!(target: "front", window = %label, "{message}"),
    }
}

#[cfg(test)]
#[path = "log.test.rs"]
mod tests;
